### PR TITLE
RFC Semi-structured Data Types

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Add RFC for Semi-structured data types design

## Changelog

- Documentation

## Related Issues
#3916

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+#### Store data in one column as JSON or binary JSON-like format
+
+JSON (JavaScript Object Notation) is the most common semi-structured format that can represent arbitrarily complex hierarchical values. It is very suitable for representing this kind of semi-structured data. Data of type `Variant` can be encoded in JSON format and stored as a raw string value.
+The main disadvantage of the JSON format is that each access requires expensive parsing of the raw string, so there are several optimized binary JSON-like formats to improve parsing speed and single key access.
+For example, MongoDB and PostgreSQL use [BSON](https://bsonspec.org/) and [jsonb](https://www.postgresql.org/docs/14/datatype-json.html) respectively to store data in JSON format.
+[UBJSON](https://ubjson.org/) is also a compatible format specification for binary JSON, it can provide universal compatibility, as easy of use as JSON while being faster and more efficient.
+All of these binary JSON formats have better performance, the only problem is they lack a good Rust implementation libraries.
```
Comment (ZhiHanZ) [inline docs/dev/rfcs/query/0007-semi-structured-data-types.md:24]: for further storage layout optimization, I think this paper also provide some insights, https://arxiv.org/pdf/2111.11517.pdf

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+Currently `Column` is only implemented for fundamental types, custom structs or enumerations like `serde_json::Value` don't have a suitable `Column` implementation to store.
+Define `ObjectColumn` and `MutableObjectColumn` as generic structures to store custom data types, and implement trait `Column` and `MutableColumn` respectively.
+`ObjectType` can be any custom type of structure or enumerations, we can define `JsonColumn` by specified parameter as `serde_json::Value`.
+All the `variant` data will be automatically cast to `serde_json::Value` and generate a `JsonColumn`.
+Other custom data types like BitmapColumn can be supported easily in the future.
+
+```rust
+#[derive(Clone)]
+pub struct ObjectColumn<T: ObjectType> {
```
Comment (sundy-li) [inline docs/dev/rfcs/query/0007-semi-structured-data-types.md:90]: ObjectType can extend from arrow2's `NativeType`.

So values can be `Arc<Bytes<T>>` which will have zero cost to take slice.

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+Semi-structured data types are used to represent schemaless data formats, such as JSON, XML, and so on.
+In order to be compatible with [Snowflake's SQL syntax](https://docs.snowflake.com/en/sql-reference/data-types-semistructured.html), we support the following three semi-structured data types:
+
+- `Variant`: A tagged universal type, which can store values of any other type, including `Object` and `Array`.
+- `Object`: Used to represent collections of key-value pairs, where the key is a non-empty string, and the value is a value of `Variant` type.
+- `Array`: Used to represent dense or sparse arrays of arbitrary size, where the index is a non-negative integer (up to 2^31-1), and values are `Variant` types.
+
+Since `Object` and `Array` can be regarded as a type of `Variant`, the following introduction mainly takes `Variant` as an example.
+
```
Comment (Xuanwo) [inline docs/dev/rfcs/query/0007-semi-structured-data-types.md:13]: Can you give an example of how users will use this feature? Like presenting a real SQL?


Take https://github.com/datafuselabs/opendal/blob/main/docs/rfcs/0000-example.md#guide-level-explanation for a look.

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+The data type of each sub-column can also be automatically deducted from the value, then we can automatically create those sub-columns and insert the corresponding value.
+
+The main advantage of this storage format is that it does not need to parse the raw JSON string when querying the data, which can greatly speed up the query processing.
+The disadvantage is that additional processing is required when inserting data, and the schema of data in each row is not exactly the same. In some scenarios with large differences, many sub-column data will be Null.
+In order to have good performance and balance in various scenarios, we can refer to the optimization algorithms introduced in the paper [JSON Tiles](https://db.in.tum.de/people/sites/durner/papers/json-tiles-sigmod21.pdf).
+
+From the perspective of performance, a better solution is to store data in binary JSON-like format and extract some frequently queried unique keys as sub-columns.
+However, in order to simplify development, we use the JSON format in the first version.
+Binary JSON-like format and separately stored sub-columns will be adopted in a future optimized version.
```
Comment (dantengsky) [inline docs/dev/rfcs/query/0007-semi-structured-data-types.md:53]: Does it mean that for the first version, those cols will not be stored in a "column-oriented" way?

how about the [dremel](https://github.com/julienledem/redelm/wiki/The-striping-and-assembly-algorithms-from-the-Dremel-paper) style encoding of nested data structure (which is supported by parquet out of the box)

Comment (sundy-li) [reply docs/dev/rfcs/query/0007-semi-structured-data-types.md:53]: That's pretty complicated, and it seems to only work for the fixed schema column.

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md (hunk 1 above)
Comment (b41sh) [reply docs/dev/rfcs/query/0007-semi-structured-data-types.md:24]: ok, I'll look into this paper

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md (hunk 2 above)
Comment (b41sh) [reply docs/dev/rfcs/query/0007-semi-structured-data-types.md:90]: ok

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md (hunk 4 above)
Comment (b41sh) [reply docs/dev/rfcs/query/0007-semi-structured-data-types.md:53]: parquet nested data type must define schema, not suitable for storing such arbitrarily data.

File: docs/dev/rfcs/query/0007-semi-structured-data-types.md (hunk 3 above)
Comment (b41sh) [reply docs/dev/rfcs/query/0007-semi-structured-data-types.md:13]: I have added some examples, PTAL @Xuanwo

**Pull request merged**